app.go

```diff
@@ -104,252 +186,15 @@ func main() {
-	session, err := mgo.DialWithTimeout(connStr, 0)
-	if err != nil {
-		return err
-	}
-	defer session.Close()
-
-	err = clearCollection(session, database, collection)
-	if err != nil {
-		return err
-	}
-
-	start := time.Now()
-	log.Printf("starting restore of %s/%s\n", database, collection)
-
-	bulk := session.DB(database).C(collection).Bulk()
-
-	var batchBytes int
-	batchStart := time.Now()
-
-	// set rate limit to 250ms
-	limiter := rate.NewLimiter(rate.Every(250 * time.Millisecond), 1)
-
-	for {
-		
-		next, err := readNextBSON(reader)
-		if err != nil {
-			return err
-		}
-		if next == nil {
-			break
-		}
-
-		// If we have something to write and the next doc would push the batch over
-		// the limit, write the batch out now. 15000000 is intended to be within the
-		// expected 16MB limit
-		if batchBytes > 0 && batchBytes+len(next) > 15000000 {
-			_, err = bulk.Run()
-			if err != nil {
-				return err
-			}
-
-			var duration = time.Since(batchStart)
-			log.Infof("Written bulk restore batch for %s/%s. Took %v", database, collection, duration)
-
-			// rate limit between writes to prevent overloading MongoDB
-			limiter.Wait(context.Background())
-
-			bulk = session.DB(database).C(collection).Bulk()
-			batchBytes = 0
-			batchStart = time.Now()
-		}
-
-		bulk.Insert(bson.Raw{Data: next})
-
-		batchBytes += len(next)
-	}
-	_, err = bulk.Run()
-	log.Printf("finished restore of %s/%s. Duration: %v\n", database, collection, time.Since(start))
-	return err
-
-}
-
-func readNextBSON(reader io.Reader) ([]byte, error) {
-	var lenBytes [4]byte
-
-	_, err := io.ReadFull(reader, lenBytes[:])
-	if err != nil {
-		if err != io.EOF {
-			return nil, err
-		}
-		return nil, nil
-	}
-
-	docLen := int32(binary.LittleEndian.Uint32(lenBytes[:]))
-
-	if docLen < 5 {
-		return nil, fmt.Errorf("invalid document size: %v bytes", docLen)
-	}
-
-	buf := make([]byte, docLen)
-	copy(buf, lenBytes[:])
-
-	_, err = io.ReadAtLeast(reader, buf[4:], int(docLen-4))
-	if err != nil {
-		if err == io.EOF {
-			// this is a broken document.
-			return nil, io.ErrUnexpectedEOF
-		}
-		return nil, err
-	}
-	return buf, nil
-}
-
-func clearCollection(session *mgo.Session, database, collection string) error {
-	start := time.Now()
-	log.Printf("clearing collection %s/%s\n", database, collection)
-	_, err := session.DB(database).C(collection).RemoveAll(nil)
-	log.Printf("finished clearing collection %s/%s. Duration : %v\n", database, collection, time.Now().Sub(start))
-
-	return err
-}
-
-type collName struct {
-	database   string
-	collection string
-}
-
-func parseCollections(colls string) ([]collName, error) {
-	var cn []collName
+func parseCollections(colls string) ([]dbColl, error) {
+	var cn []dbColl
 	for _, coll := range strings.Split(colls, ",") {
 		c := strings.Split(coll, "/")
 		if len(c) != 2 {
-			return nil, fmt.Errorf("failed to parse connections string : %s\n", colls)
+			return nil, fmt.Errorf("failed to parse connections string: %s\n", colls)
```

Comment:
I would remove the newline from this error message.

Reply:
removed.